### PR TITLE
Refactor redux

### PR DESCRIPTION
**Pull request by BeroBurny**
**Short description of work done**
Refactor Redux state to new `ducks` pattern with redux toolkit and redux sagas

### PR Checklist

- [x] I have run linter localy
- [ ] I have run unit and integration tests locally
- [x] Rebased to master branch / merged master

### Changes
- implement redux toolkit and redx sagas
- rewrite logic to use redux toolkit and redx sagas
- use ducks pattern
- implement selectors

### TODO
- fix tests

### CODE REVIEW

File: src/renderer/ducks/network/actions.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import {createAction} from "@reduxjs/toolkit";
+import {DockerPort} from "../../services/docker/type";
+
+export const {
+    startDockerImagePull, endDockerImagePull,
+    selectNetwork, loadedValidatorBeaconNodes,
+    subscribeToBlockListening, unsubscribeToBlockListening
+} = networkSlice.actions;
+
+type StartBeaconChain = (network: string, ports?: DockerPort[]) => {payload: {network: string, ports?: DockerPort[]}};
```
Comment (morrigan) [inline src/renderer/ducks/network/actions.ts:11]: let's not inline types as it's really hard to read

```
type StartBeaconChain = (network: string, ports?: DockerPort[]) => {
   payload: {
    network: string,
    ports?: DockerPort[],
   },
}
```

File: src/renderer/ducks/deposit/actions.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+export const {
+    storeDepositTx, depositDetected, depositNotFount, waitForDeposit, resetDepositData
+} = depositSlice.actions;
+
+export const generateDeposit = createAction<INetworkConfig>("deposit/generateDeposit");
+
+type VerifyDeposit = (networkConfig: INetworkConfig, timeout?: number) =>
+{payload: {networkConfig: INetworkConfig; timeout: number}};
+export const verifyDeposit = createAction<VerifyDeposit>(
```
Comment (morrigan) [inline src/renderer/ducks/deposit/actions.ts:13]: same here with inline, it's hard to read. maybe we can add an eslint rule for this :)

File: src/renderer/ducks/deposit/sagas.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    const provider = networkConfig.eth1Provider;
+    const ethersNotifier = new EthersNotifier(networkConfig, provider);
+
+    // Call deposit service and listen for event, when transaction is visible dispatch action
+    // TODO: Refactor entire logic and service logic
+    try {
+        yield ethersNotifier.depositEventListener(keyPair.publicKey, timeout);
+        yield put(depositDetected());
+    } catch {
+        yield put(depositNotFount());
```
Comment (morrigan) [inline src/renderer/ducks/deposit/sagas.ts:59]: typo :P

File: src/renderer/ducks/deposit/sagas.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    // TODO note to myself: if request resolve() or reject() still check lock if is deposit made?
+    // Is there point of using logic like this? or will be better to use different approach
+    const hasDeposited: boolean = yield ethersNotifier.hasUserDeposited(keyPair.publicKey);
```
Comment (morrigan) [inline src/renderer/ducks/deposit/sagas.ts:64]: I am not sure what confuses you with this approach.

However, maybe we should check first if deposit was made and only if not, then activate the event listener?

File: src/renderer/ducks/register/slice.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            reducer: (
+                state,
+                action: PayloadAction<{signingKey: string; withdrawalKey: string; signingKeyPath: string;}>
+            ): void => {
+                state.signingKey = action.payload.signingKey;
+                state.withdrawalKey = action.payload.withdrawalKey;
+                state.signingKeyPath = action.payload.signingKeyPath;
+            },
+            prepare: (signingKey: string, withdrawalKey: string, signingKeyPath: string):
+            {payload: {signingKey: string; withdrawalKey: string; signingKeyPath: string;}} => ({
```
Comment (morrigan) [inline src/renderer/ducks/register/slice.ts:45]: export `signingKey: string; withdrawalKey: string; signingKeyPath: string;` to interface as it's reused a lot here

File: package.json [hunk 6]
```diff
@@ -196,6 +197,8 @@
     "recharts": "^1.8.5",
     "redux": "^4.0.1",
     "redux-logger": "^3.0.6",
+    "redux-saga": "^1.1.3",
+    "redux-saga-test-plan": "^4.0.1",
     "redux-thunk": "^2.3.0",
```
Comment (morrigan) [inline package.json:202]: remove if not used anymore

**Review by morrigan — COMMENTED** [review]
Idk, generally looks good! Left few comments mostly regarding code style mostly

File: src/renderer/ducks/deposit/sagas.ts (hunk 4 above)
Comment (BeroBurny) [reply src/renderer/ducks/deposit/sagas.ts:64]: forget to remove this `TODO`, we already talked about this

File: src/renderer/ducks/deposit/actions.ts (hunk 2 above)
Comment (BeroBurny) [reply src/renderer/ducks/deposit/actions.ts:16]: best solution is to implement `prettier`

File: package.json [hunk 7]
```diff
@@ -198,6 +198,7 @@
     "redux": "^4.0.1",
     "redux-logger": "^3.0.6",
     "redux-saga": "^1.1.3",
+    "redux-saga-test-plan": "^4.0.1",
```
Comment (mpetrunic) [inline package.json:201]: dev dependency?

**Review by mpetrunic — COMMENTED** [review]
LGTM. We should probably merge this and continue rest of stuff in smaller PRs as this is quite hard to review

**Review by morrigan — APPROVED** [review]
👍

**Pull request merged**